File: solutions/0599. Minimum Index Sum of Two Lists/0599.java
class Solution {
  public String[] findRestaurant(String[] list1, String[] list2) {
    List<String> ans = new LinkedList<>();
    Map<String, Integer> restaurantToIndex = new HashMap<>();
    int minSum = Integer.MAX_VALUE;

    for (int i = 0; i < list1.length; ++i)
      restaurantToIndex.put(list1[i], i);

    for (int i = 0; i < list2.length; ++i) {
      if (restaurantToIndex.containsKey(list2[i])) {
        int sum = restaurantToIndex.get(list2[i]) + i;
        if (sum < minSum)
          ans.clear();
        if (sum <= minSum) {
          ans.add(list2[i]);
          minSum = sum;
        }
      }
    }

    return ans.toArray(new String[0]);
  }
}

File: solutions/0599. Minimum Index Sum of Two Lists/0599.py
class Solution:
  def findRestaurant(self, list1: List[str], list2: List[str]) -> List[str]:
    ans = []
    restaurantToIndex = {restaurant: i for i,
                         restaurant in enumerate(list1)}
    minSum = inf

    for i, restaurant in enumerate(list2):
      if restaurant in restaurantToIndex:
        sum = restaurantToIndex[restaurant] + i
        if sum < minSum:
          ans.clear()
        if sum <= minSum:
          ans.append(restaurant)
          minSum = sum

    return ans
